open-hackathon-server/src/hackathon/azureformation/service.py

```diff
@@ -62,6 +64,11 @@ class Service(ServiceManagementService, Component):
     def __init__(self, azure_key_id):
         self.azure_key_id = azure_key_id
         azure_key = self.db.get_object(AzureKey, self.azure_key_id)
+
+        if not isfile(azure_key.pem_url):
+            cryptor = RequiredFeature("cryptor")
+            cryptor.recover_local_file(azure_key.azure_pem_url, azure_key.pem_url)
```

Comment:
I think it's better not include those logic in `azureformation` logic. The `Service` class can access a valid pem_url as parameter to keep `azureformation` clean.
Also that `azureformation` was refactored by Qi Dong. The whole module will be removed and all classed will be moved to `hazure` module. To keep azure classes clean is much easier for the merge work